### ISSUE TITLE
Implement numpy.clip()

### ISSUE THREAD

**Issue by certik** [lcompilers/lpython#334]
This is a good function to implement fully in LPython. It works for:

* any type
* any array rank and some combination of ranks
* it automatically broadcasts

If we can fully implement this as regular LPython code, without any extra "`clip` specific" logic in AST2ASR, this will be enough to implement large parts of NumPy.

So we can use this function as a "testbed", and implement all missing features in LPython.

I suggest we try to implement some small subset, then we try to make the function work for more combination of types and ranks, then we try to implement some generalizations in LPython to make this easier, such as "generics" (any type) and rank agnostic generics, to make the code simpler, and iteratively improve, until this function can be implemented relatively easily.

See also:

* NumBa implementation: https://github.com/numba/numba/pull/7958

**Comment by czgdp1807** [comment]
Some of the high priority features to be added before `np.clip` can be implemented,

1. Generics - We need overloaded definitions for `np.clip` for different cases like because, `a` belongs to `{scalar, array}`, `a_min` belongs to `{scalar, array, NoneType}` and `a_max` also belongs to `{scalar, array, NoneType}`. So, there will be 18 different cases for these three arguments.
2. Broadcasting arrays - In case of arrays, ranks can be different and we might need to broadcast all the input arrays to same shape. This can be achieved via explicit broadcasting i.e., by actually copying the data and extend the dimensions wherever needed. Or, implicit broadcasting i.e., by generating indices to fetch the right value from each array. For example, in case of `(5, 2, 4)` and `(2, 4)`, for `(i, 1, 2)` (0 <= i < 4) you need to fetch `(1, 2)` from the second array.

**Comment by certik** [comment]
I would suggest we work on all these fronts at once, so start on generics and broadcasting, and at the same time try to create the best (but incomplete) version of `clip` using existing features of LPython.

**Comment by czgdp1807** [comment]
Yes. We can do that because, both generics and broadcasting are independent of each other and can be implemented simultaneously. May be let's start by taking a simple example for each of these and update LPython so that it can compile these examples. Then we can go for more advanced cases.
